pandoras_flask/metrics.py

```diff
@@ -0,0 +1,57 @@
+import prometheus_client
+import time
+
+from flask import request, Response
+from prometheus_client import multiprocess, Counter, Histogram
+
+# Multiprocessing setup
+# Cf. https://github.com/prometheus/client_python#multiprocess-mode-gunicorn
+registry = prometheus_client.CollectorRegistry()
+multiprocess.MultiProcessCollector(registry)
+
+# Metrics for the `app` module - reads & writes
+
+PING_RESPONSE = Counter('ping_response',
+                        '/ping response codes', ['code'])
+
+# Middleware / setup
+# Following the Flask multiprocessing example in
+# https://github.com/amitsaha/python-prometheus-demo
+
+REQUEST_COUNT = Counter('request_count', 'App Request Count',
+                        ['method', 'endpoint', 'http_status'])
+
+REQUEST_LATENCY = Histogram('request_latency_seconds', 'Request Latency',
+                            ['endpoint'])
+
+
+def start_timer():
+    request.start_time = time.time()
+
+
+def stop_timer(response):
+    resp_time = time.time() - request.start_time
+    REQUEST_LATENCY.labels(request.path).observe(resp_time)
+    return response
+
+
+def record_request_data(response):
+    REQUEST_COUNT.labels(request.method, request.path,
+                         response.status_code).inc()
+    return response
+
+
+# Cf. Prometheus exposition format in https://git.io/fpSOY
+CONTENT_TYPE_LATEST = str('text/plain; version=0.0.4; charset=utf-8')
+
+
+def setup(app):
+    app.before_request(start_timer)
+    # The order here matters since we want stop_timer to be executed first
```

Comment:
The message here states that `stop_timer` is to be executed first but then `record_request_data` is the actual first method to be executed in the code below.

I feel this may be a typo or that the ordering doesn't matter?

Reply:
This is from the code we followed - https://github.com/amitsaha/python-prometheus-demo/blob/master/flask_app_prometheus_multiprocessing/src/helpers/middleware.py#L39

The documentation for [after_request](http://flask.pocoo.org/docs/1.0/api/#flask.Flask.after_request) suggests there's no ordering at all, which is not great. :thinking:

Looking at the code, it *is* ordered:
https://github.com/pallets/flask/blob/a74864ec229141784374f1998324d2cbac837295/flask/app.py#L2125

The list of functions is appended to when you add one, and they're executed in reverse order.

I'll update the comment to make that clear.

Reply:
Ah, reverse order execution strikes again :) Thanks for clearing that up. I'll approve the PR in lieu of the two comments to be added. Thanks!